File: sonaranalyzer-dotnet/src/Tests/SonarAnalyzer.UnitTest/TestCases/AvoidUnsealedAttributes.cs
using System;

namespace Tests.Diagnostics
{
    public class MyAttribute : Attribute // Noncompliant {{Seal this attribute or make it abstract.}}
//               ^^^^^^^^^^^
    {
    }

    protected class MyOtherAttribute : Attribute // Noncompliant
    {
    }

    protected internal class MyOtherAttribute2 : Attribute // Noncompliant
    {
    }

    public sealed class Bar : Attribute // Compliant - sealed
    { }

    public abstract class FooBar : Attribute // Compliant - abstract
    {
    }

    public sealed class Attr : Attribute
    {
        private class InnerAttr : Attribute // Compliant - private
        {
            public class InnerInnerAttr : Attribute // Compliant - effective accessibility is private
            {
            }
        }
    }
}

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/AvoidUnsealedAttributes.vb
Public Class PublicAttribute ' Noncompliant {{Seal this attribute or make it abstract.}}
    Inherits Attribute
    '        ^^^^^^^^^^^^^^^ @-1
End Class

Public NotInheritable Class SealedAttribute
    Inherits Attribute ' Compliant
End Class

Public MustInherit Class AbstractAttribute
    Inherits Attribute ' Compliant
End Class

Public Class Container

    Protected Class ProtectedAttribute ' Noncompliant
        Inherits Attribute
    End Class

    Private Class PrivateAttribute ' Compliant
        Inherits Attribute

        Public Class SubClassAttribute ' Compliant - effective accessibility is private
            Inherits Attribute
        End Class
    End Class
End Class
